store/primary/primary.go

```diff
@@ -126,38 +173,20 @@ func duplicateEntry(in store.IndexEntry, old []store.IndexEntry) bool {
 	return false
 }
 
-// checksize to see if we need to create a new tree
-// If the size is equal or over sizeLimit create new instance
-// of the tree
-// NOTE: Should we add a security margin? Aim for the 0.45 instead
-// of 0.5
-func (s *rtStorage) checkSize() {
-	// s.fullLock()
-	// defer s.fullUnlock()
-	s.primary.lk.RLock()
-	// TODO: Here we are just looking to the length of the radix tree not the
-	// actual storage size. This needs to be changed so we check the actual size being
-	// used by the tree.
-	if s.primary.tree.Len() >= int(float64(0.5)*float64(s.sizeLimit)) {
-		s.lk.Lock()
-		// Create new tree and make primary = secondary
-		s.secondary = s.primary
-		s.primary = &tree{tree: radixtree.New()}
-		s.lk.Unlock()
+func (s *rtStorage) lockBranch(k string) {
+	var idx int
+	if k != "" {
+		// bitwise modulus requires that s.locks is power of 2
+		idx = int(k[0]) & (len(s.branchLocks) - 1)
```

Comment:
Wouldn't it be better to use `k[len(k)-1]` instead of `k[0]` here? Multihashes use prefixes that may be biasing the lock that we lock, right? (see https://github.com/multiformats/multihash).

Maybe I am wrong, is just a sanity-check.

Reply:
That would be better, but...  We need to lock a "branch" of the tree so that one lock prevents any operation in the same section of the tree, so we need to use the **first** bits of the key.  There is an implicit relationship between this locking strategy and how a radix tree works.

That being said, there is a way to solve both problems at the same time: Store the key backwards.

Reply:
Oh, now I get it! Really smart! 

I don't think we need to store the key backwards right now, but worth having it in mind (adding a NOTE or something?) in case we start seeing some performance overhead as a result of using the multihash prefix for the lock.

Reply:
I think we want to store the key forwards because that automatically compresses out any common prefix.  Another way to solve this is to use multiple trees, and compute a fast hash of the key to see which tree to use.  Then just lock that tree. I will revisit this problem.

Reply:
Added comment